### DISCUSSION TITLE
Long Date Format (Tue November 14, 2023)

### DISCUSSION THREAD

**Discussion by iicurtis** [pysan3/neorg-templates#15] [Ask for a Snippet]
Thanks for making this project! I think it's a great use of an existing tool to extend another one. 

Do you happen to have a snippet that creates the current date in a long written format? I am hoping for something similar to `TODAY_OF_FILENAME` but with a different end format. Unfortunately, I don't know where the documentation is for the format function in this snippet. Does it support day of week and month of year?

Example: `Tue November 14, 2023` or `Tue 14 November 2023`

**Comment by pysan3** [comment]
Thanks for your interest!

I'm currently on my phone so I'll definitely give you a long answer later but let me ask a few things. 

`TODAY_OF_FILENAME` is meant to extract the datetime from the filename generated by journal module. So it means January 1st, 2023 when opening 2023-01-01.norg and not today's date. (FYI, use `TODAY_OF_FILETREE` for 2023/01/01.norg)

Do you want this functionality but the output to be long date format? Or a simple long date of today, meaning literally the date that's invoked?

As a sidenote, parsing file names written in the long date format would be quite a pain and I'm hoping that's not what you want lol.

**Reply by iicurtis** [reply]
I would like to parse the date from the filename (like `TODAY_OF_FILENAME` currently does), but I can probably connect the dots if you can help me figure out the formatting. 

> Do you want this functionality but the output to be long date format? Or a simple long date of today, meaning literally the date that's invoked?

Yes. I would like the snippet to generate a long date (Sat April 01, 2023) for the file name (ex. 2023-04-01.norg). I tend to have journal entries that I type up from written notes that occurred before the current day. Eventually, I would probably also try to build the long date of current date too, but that's a lower priority goal. 

I do not want to parse long-date filenames, that would be awful lol.

**Comment by iicurtis** [comment]
Other common formats of interest might be:

- org mode standard: `<2006-11-01 Wed 19:15>`
- norg mode standard: `<day>?,? <day-of-month> <month> -?<year> <time> <timezone>` ex: `Wed, 1 Nov 2006 19:15`

**Comment by pysan3** [comment]
Hi @iicurtis , as of v2.0.1, you now have the following options.

1. Control globally with plugin option.

As I stated here: https://github.com/pysan3/neorg-templates#snippets_overwrite, there is a global option to specify the output format of all `TODAY_*, YESTERDAY_*, ...` snippets. (I found a bug here so glad you asked.)
You may use lua time tags (`%d, %x, ...`) [option list](https://www.lua.org/pil/22.1.html) here, when you scroll down a bit, you'll see a table.

2. Build you own snippet key.

If you don't want to alter the default behavior, but still want to use a different style occasionally, I refactored the default snippets so that you can specify the date format individually.
You can use `parse_date(delta_date, date, date_format)` where `delta_date` shifts the date (-1 for yesterday) and `date_format` is the format string mentioned above. You will likely want to pass `file_name_date()` or `file_tree_date()` as the second argument.

More details here: https://github.com/pysan3/neorg-templates#keywords.

```lua
neorg.setup({
    -- ...
    ["external.templates"] = {
      config = {
        keywords = {
          TODAY_OF_ORG = function() -- detect date from filename and return in org date format
            local ls = require("luasnip")
            local s = require("neorg.modules.external.templates.default_snippets")
            return ls.text_node(s.parse_date(0, s.file_name_date(), [[%Y-%m-%d %a]])) -- 2006-11-01 Wed
          end,
          NOW_IN_DATETIME = function() -- print current date+time of invoke
            local ls = require("luasnip")
            local s = require("neorg.modules.external.templates.default_snippets")
            return ls.text_node(s.parse_date(0, os.time(), [[%Y-%m-%d %a %X]])) -- 2023-11-01 Wed 23:48:10
          end,

        }
      },
    },
}

**Reply by pysan3** [reply]
@iicurtis does this work?

**Reply by iicurtis** [reply]
Thank you! This is really awesome. With this config, I can get the keywords working great! 

Although my problem is solved, I was unable to get the global config option working for some reason. The `snippets_overwrite` does nothing in the below example. Using keywords `TODAY` and `TODAY_OF_FILE` both resulted in the default `2023-11-21` format.

```lua
        ["external.templates"] = {
          config = {
            snippets_overwrite = {
              date_format = [[%a %d]],
            },
            keywords = {
              TODAY_OF_FILE_ORG = function() -- detect date from filename and return in org date format
                local ls = require("luasnip")
                local s = require("neorg.modules.external.templates.default_snippets")
                return ls.text_node(s.parse_date(0, s.file_name_date(), [[<%Y-%m-%d %a]])) -- <2006-11-01 Wed>
              end,
              TODAY_OF_FILE_NORG = function() -- detect date from filename and return in norg date format
                local ls = require("luasnip")
                local s = require("neorg.modules.external.templates.default_snippets")
                return ls.text_node(s.parse_date(0, s.file_name_date(), [[%a, %d %b %Y]])) -- Wed, 1 Nov 2006
              end,
              TODAY_ORG = function() -- detect date from filename and return in org date format
                local ls = require("luasnip")
                local s = require("neorg.modules.external.templates.default_snippets")
                return ls.text_node(s.parse_date(0, os.time(), [[<%Y-%m-%d %a %H:%M:%S>]])) -- <2006-11-01 Wed 19:15>
              end,
              TODAY_NORG = function() -- detect date from filename and return in norg date format
                local ls = require("luasnip")
                local s = require("neorg.modules.external.templates.default_snippets")
                return ls.text_node(s.parse_date(0, os.time(), [[%a, %d %b %Y %H:%M:%S]])) -- Wed, 1 Nov 2006 19:15
              end,
            },
          },
        },

```

**Reply by pysan3** [reply]
> I was unable to get the global config option working for some reason. The snippets_overwrite does nothing in the below example.

You are right man, Sorry I had a bug in the code!
